lexer.py

```diff
@@ -1,69 +1,111 @@
-class Token:
-    def __init__(self, type, value=None):
-        self.type = type
-        self.value = value
+from tokens import Token, TokenType
+from errors import InvalidCharacterException, NoValidTokenException
 
-    def __repr__(self):
-        # This can be reverted back to f' format. Same thing.
-        return 'Token({type}, {value})'.format(
-            type=self.type,
-            value=repr(self.value)
-        )
 
 class Lexer:
-    def __init__(self, text):
+    def __init__(self, text: str):
         self.text = text
         self.pos = 0
         self.current_char = self.text[self.pos]
         self.tokens = []
+        self.lineCount = 1
 
-    def error(self):
-        raise Exception('Invalid character')
-
-    def advance(self):
-        self.pos += 1
+    def advance(self, by: int = 1):
+        self.pos += by
         if self.pos < len(self.text):
             self.current_char = self.text[self.pos]
         else:
             self.current_char = None
 
-    def integer(self):
+    '''
+    DOCSTRING: The number() function is recursive and has 4 parts:
+        1. While the current_char is a digit, it will be appended to 'result'
+        2. If a decimal point is enountered, it is appended along with a recursive call to number()
+        3. If the result contains a decimal point, the float number is returned
+        4. Otherwise, it is either an integer, or the numbers past the decimal point in one of the recursions. Either way
+            the int value can be returned.
+    The else block contains error handling for the edge case of an input like "44.hi" which would raise a NoValidTokenError.
+    All errors are caught and returned in the tokenize method.
+    '''
+    def number(self):
         result = ''
         while self.current_char is not None and self.current_char.isdigit():
             result += self.current_char
             self.advance()
```

Comment:
Everything looks great and I was able to follow it during my debug session. My only suggestion is possibly refactoring the assignment and advance into one function.

ie

```
def consumeAdvance(self):
    char = self.current_char
    self.advance()
    return char

then you can do
result += self.consumeAdvance()
    result += self.current_char
    self.advance()
```
```suggestion
            result += self.current_char
            self.advance()
```

Reply:
I like this idea! Instead of defining a new function, could we just add a return statement to the already defined `advance()` method? Then in cases where we need the character (like here), we can use the return value, and in other cases, we can just call the method and not assign it to anything. Thoughts?